### DISCUSSION TITLE
asycnpg + ORM performance issue (pre-buffered ORM results vs. non)

### DISCUSSION THREAD

**Discussion by ludaavics** [sqlalchemy/sqlalchemy#7294] [Usage Questions]
### Describe the bug

 - ORM with asyncpg driver is an order of magnitude slower than with sync drive ..
 - ... even though using asyncpg directly is faster than using psycopg2 directly
 
I guess there is an overhead to async but an order of magnitude seems a bit much? Furthermore, the fact that performance issue is tied to the ORM+async suggest something specific to sqlalchemy implementation? 

Is this expected?

### To Reproduce

```python
import asyncio
import asyncpg
import psycopg2

import random
import time
import sqlalchemy as sa
from sqlalchemy import Column, Integer, Float
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine

Base = declarative_base()

user = "user"
password = "password"
host = "host"
port = "port"
db = "db"
schema = "schema"
table = "test"
n_rows = 1_000_000
n_cols = 5


async_engine = create_async_engine(
    f"postgresql+asyncpg://{user}:{password}@{host}:{port}/{db}",
    echo=False,
)
async_session = sessionmaker(async_engine, class_=AsyncSession)

sync_engine = sa.create_engine(
    f"postgresql://{user}:{password}@{host}:{port}/{db}",
    echo=False,
)
sync_session = sessionmaker(sync_engine, expire_on_commit=False)


class Test(Base):
    __tablename__ = table
    id = Column(Integer, primary_key=True)
    n1 = Column(Float)
    n2 = Column(Float)
    n3 = Column(Float)
    n4 = Column(Float)
    n5 = Column(Float)

    __table_args__ = {"schema": schema}


async def async_main():
    async with async_engine.begin() as async_conn:
        await async_conn.run_sync(Base.metadata.drop_all)
        await async_conn.run_sync(Base.metadata.create_all)

        data = [
            {f"n{i + 1}": random.random() for i in range(n_cols)} for j in range(n_rows)
        ]

        await async_conn.execute(
            sa.text(
                f"INSERT INTO {schema}.{table} (n1, n2, n3, n4, n5) VALUES (:n1, :n2, :n3, :n4, :n5)"
            ),
            data,
        )
        await async_conn.commit()

    async_conn = await asyncpg.connect(
        user=user,
        password=password,
        database=db,
        host=host,
        port=port,
    )

    sync_conn = psycopg2.connect(
        f"postgresql://{user}:{password}@{host}:{port}/{db}"
    ).cursor()

    raw_sql = f"SELECT * FROM {schema}.{table}"

    start = time.time()
    await async_conn.fetch(raw_sql)
    end = time.time()
    print(f"Raw asyncpg took {end - start:.3f}s")

    start = time.time()
    sync_conn.execute(raw_sql)
    end = time.time()
    print(f"Raw psycopg2 took {end - start:.3f}s")

    async with async_session() as session:
        start = time.time()
        await session.execute(sa.text(raw_sql))
        end = time.time()
        print(f"Sqlalchemy async raw took {end - start:.3f}s")

        start = time.time()
        await session.execute(sa.select(Test))
        end = time.time()
        print(f"Sqlalchemy async ORM took {end - start:.3f}s")

    with sync_session() as session:
        start = time.time()
        session.execute(sa.text(raw_sql))
        end = time.time()
        print(f"Sqlalchemy sync raw took {end - start:.3f}s")

        start = time.time()
        session.execute(sa.select(Test))
        end = time.time()
        print(f"Sqlalchemy sync ORM took {end - start:.3f}s")


if __name__ == "__main__":
    asyncio.run(async_main())
```

### Error

```
Raw asyncpg took 1.755s
Raw psycopg2 took 2.396s
Sqlalchemy async raw took 1.652s
Sqlalchemy async ORM took 14.662s
Sqlalchemy sync raw took 2.733s
Sqlalchemy sync ORM took 2.289s
``` 

### Versions

- OS: ubuntu 18.04
- Python: 3.6.10
- SQLAlchemy: 1.4.11
- Database: postgresql 9.6
- DBAPI pscopg2 / asyncpg

### Additional context

As an added note, in my real data, even raw SQL execution is about 3x slower using sqlalchemy with asyncpg driver vs using asyncpg (or psycopg2) directly. I obviously haven't been able to replicate that with mock data, so maybe something idiosyncratic to my setup, but i thought i would highlight.

**Comment by zzzeek** [comment]
if im reading correrctly the main issue with your test is you are not fetching the rows back.  that's where all the time is spent, and the asyncio extension necessarily buffers the ORM objects generated within the execute() call, as you are getting back a sync Result object.  if you pass the prebuffer_rows execution option to the ORM session .execute as in the example at https://docs.sqlalchemy.org/en/14/errors.html?highlight=prebuffer_rows#object-cannot-be-converted-to-persistent-state-as-this-identity-map-is-no-longer-valid you will get the same performance for a SELECT.

there's also a dramatic performance difference in psycopg2 vs. asyncpg on the INSERT side due to the [fast execution helpers](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#psycopg2-executemany-mode) which we have not yet implemented for asyncpg.

**Comment by ludaavics** [comment]
First off, thanks for the consistently blazing fast answers!

OK that makes sense that I wasn't comparing apples to apples. Indeed if I fetch the rows in the sync driver it ends up again slower than async driver. I'm glad I asked, I was about to just go back to using psycopg2.
```python
async with async_session() as session:
    start = time.time()
    result = await session.execute(sa.select(Test))
    data = result.scalars().all()
    end = time.time()
    print(f"Sqlalchemy async ORM took {end - start:.3f}s")  # 18s

with sync_session() as session:
    start = time.time()
    result = session.execute(sa.select(Test))
    data = result.scalars().all()
    end = time.time()
    print(f"Sqlalchemy sync ORM took {end - start:.3f}s") # 27s
```

Also, when i select the columns directly, rather than the full object, I get much better performance -- I'd never realized just how much more efficient not building the object was, thanks!

```python
async with async_session() as session:
    start = time.time()
    result = await session.execute(sa.select([Test.id, Test.n1, Test.n2, Test.n3, Test.n4, Test.n5]))
    data = result.scalars().all()
    end = time.time()
    print(f"Sqlalchemy async ORM took {end - start:.3f}s") 

with sync_session() as session:
    start = time.time()
    result = session.execute(sa.select([Test.id, Test.n1, Test.n2, Test.n3, Test.n4, Test.n5]))
    data = result.scalars().all()
    end = time.time()
    print(f"Sqlalchemy sync ORM took {end - start:.3f}s")
    
# output
Raw asyncpg took 0.919s
Raw psycopg2 took 2.427s
Sqlalchemy async raw took 1.605s
Sqlalchemy async ORM took 3.249s
Sqlalchemy sync raw took 2.857s
Sqlalchemy sync ORM took 5.421s
```

Finally, if I understand your link correctly, the reason for async session's buffer is to allow related objects to load? If so does it make sense for the async session to pre-fetch an object that doesn't have any relationship (or if it has relationship configured with raiseload)?

**Reply by filippaquant** [reply]
Hey @ludaavics , @zzzeek 
Can you share how you managed to solve this at the end?
I just faced the same issue, but looks like adding `execution_options={"prebuffer_rows": True}` to `execute()` doesn't increase the speed in case of simple select. Do I miss something ?

**Comment by zzzeek** [comment]
> Finally, if I understand your link correctly, the reason for async session's buffer is to allow related objects to load? If so does it make sense for the async session to pre-fetch an object that doesn't have any relationship (or if it has relationship configured with raiseload)?

that's the main reason.     having pre_buffer be conditional would mean just that, we have to figure out based on the compiled form of the construct (which is where decisions like whether or not to eagerload) whether or not pre_buffer is appropriate and then set it.  this is actually not that hard as we already have other decisions that are made like this for ORM queries, but it is one more conditoinal behavior which means we have to a. impement b. test c. make sure remains consistent on new kinds of cases and of course d. explain in the documentation, which is the hardest part.   a story where "asyncio result sets pre_buffer by default, you can change it like this" is IMO easy to explain.

beyond that, the notion of results being buffered within the execute() call is mostly the norm and not the exception; most DBAPIs will fully buffer the results when you call cursor.execute() unless you are explicitly using a server side cursor.   So in that sense it would actually be better for us to add pre_buffer to the sync ORM call, that way all systems would be fully consistent.   Right now, the sync session has a bit of a gap to jump over, where the DBAPI will load all the rows when we say cursor.execute(), but the ORM hasn't yet gone through the effort to iterate through those rows and interpret them as ORM objects.     so really, both systems "buffer" the result set it's just the sync ORM currently leaves half of the "buffering" until after execute() is called.  and, even more difficult to explain in a simple story, the ORM by default will run through the whole result set when you just fetch the first row anyway, like the moment you say fetchone(), it processes all the rows and buffers them.   so in that sense, setting pre_buffer default to true seems like it would make the whole story easier to explain (so far, we havent really had to explain this).

if you are using yield_per or the async stream() method, then nothing is buffered, all of the above doesn't apply.

**Comment by zzzeek** [comment]
moving this to a discussion
